Tutorial/app/src/main/java/ably/io/tutorial/ExampleActivity.java

```diff
@@ -21,6 +35,68 @@ protected void onCreate(Bundle savedInstanceState) {
         } catch (AblyException e) {
             e.printStackTrace();
         }
+        etUsername = (EditText) findViewById(R.id.editText);
+        btLogin = (Button) findViewById(R.id.button);
+        btLogin.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                new Thread(new Runnable() {
+                    @Override
+                    public void run() {
+                        /* change UI depending on the state */
+                        if (isUserLoggedIn) {
+                            logout();
+                        } else {
+                            login();
+                        }
+                    }
+                }).start();
+            }
+        });
+    }
+
+    private void login() {
+        RequestQueue queue = Volley.newRequestQueue(this);
```

Comment:
This feels like a very odd approach to me, but I'm interested to know why you chose this approach and if it is perhaps normal in an Android context.  Generally, from my experience, an app provides a means to authenticate.  It will send a request to a server at the time of the login to confirm the credentials are correct, and perhaps obtain a token such as JWT.  This then allows the app itself to be logged in on the device, but that state is not linked to the server state.  So if the app is shut down and re-opened, it can still be logged in as the login state is on the device, not on the server.  Therefore, if in turn you send a request to your server, you can present a token of some sort or even the full credentials to authenticate with each request to the sever.  It seems you've effectively wrapped an HTTP library that maintains cookies, but then the app's state itself can easily fall out of sync with the server, and relying on cookies in an app, as far as I aware, is not normal.

The purpose of this demo was to demonstrate that the client library, when it needs a token, will call `authCallback`, and the developer using a callback will then have the logic to send a request to their server to obtain a token. As this demo does not really support true authentication, the authCallback would then simply send the username without a password to obtain a token.  I am worried that this pattern in this tutorial is encouraging people to think the wrong way.